File: lib/domain/epub_reader/entities/epub_book.dart
import 'dart:typed_data';

import 'package:equatable/equatable.dart';
import 'package:leafy/domain/epub_reader/entities/epub_chapter.dart';
import 'package:leafy/domain/epub_reader/entities/epub_image.dart';

class EpubBook extends Equatable {
  final String fileName;
  final String title;
  final String author;
  final String language;
  final Uint8List? coverImage;
  final List<EpubChapter> chapters;
  final List<EpubImage> images;

  const EpubBook({
    required this.fileName,
    required this.title,
    required this.author,
    required this.language,
    this.coverImage,
    required this.chapters,
    required this.images,
  });

  @override
  List<Object?> get props => [
    fileName,
    title,
    author,
    language,
    coverImage,
    chapters,
  ];
}

File: android/app/src/main/kotlin/com/nmheeir/leafy/epub/models/EpubBook.kt
package com.nmheeir.leafy.epub.models

import android.graphics.Bitmap
import com.nmheeir.leafy.epub.cache.BitmapSerializer
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import kotlinx.serialization.protobuf.ProtoNumber

/**
 * Represents an epub book.
 *
 * @param fileName The name of the epub file.
 * @param title The title of the book.
 * @param author The author of the book.
 * @param language The language code of the book.
 * @param coverImage The cover image of the book.
 * @param chapters The list of chapters in the book.
 * @param images The list of images in the book.
 */
@OptIn(ExperimentalSerializationApi::class)
@Serializable
data class EpubBook @OptIn(ExperimentalSerializationApi::class) constructor(
    @ProtoNumber(1) val fileName: String,
    @ProtoNumber(2) val title: String,
    @ProtoNumber(3) val author: String,
    @ProtoNumber(4) val language: String,
    @ProtoNumber(5) @Serializable(with = BitmapSerializer::class) val coverImage: Bitmap?,
    @ProtoNumber(6) val chapters: List<EpubChapter> = emptyList(),
    @ProtoNumber(7) val images: List<EpubImage> = emptyList()
)